### ISSUE TITLE
iOS (16 & 17) Moves the whole canvas with all the objects Up on iPhones that have Notch

### ISSUE THREAD

**Issue by Tarek200** [Scirra/Construct-bugs#7626]
## Problem description

The game starts with everything working perfectly so the problem is: 
When we close the APP it sends it to the Background pausing the App and if we reopen it again later on it will break the App by shifting the whole canvas up, you can see all the UI, BG  and everything on the canvas moved up leaving a small gap at the bottom of the canvas.
This happens only on phones with Notchs like XS Max
For example, it doesn't happen on iPhone 8 Plus as it doesn't have a notch so that's why I suspect this bug is only related to phones with notch.

## Attach a .c3p

[iOS 16 Notch BUG.zip](https://github.com/Scirra/Construct-bugs/files/13599654/iOS.16.Notch.BUG.zip)

## Steps to reproduce

1. Export the project as (iOS Cordova)
2. Install the App into an iPhone with Notch using Xcode
3. Run the App on the phone 
4. Close it by sending the App to the Background and leave it for some time
5. Open the App again and check if the canvas moved up
6. If is still working normally then close it again and leave it some time, keep repeating all the steps from (1 to 6) until it fails
7. Note:
The bug doesn't happen right away so you may try on the first try and you will not be able to reproduce it, so please keep trying for one or two days as sometimes happens after it has been paused in the  Background for hours. We dont know the exact way that this happens as it happens randomly. Every time you test and you cannot reproduce it you need to close it back and leave it a bit longer in the background until eventually it fails. We have been reproducing this bug all the time, it is just the time that takes to fail varies between each fail. 
The only way to put the canvas back in the right position is to completely close the App by restarting the iPhone.

## Observed result

The whole Canvas moves up with all the objects 

![002](https://github.com/Scirra/Construct-bugs/assets/41597700/c3745a53-2afe-4992-9e21-738b3e31693c)

## Expected result

The canvas and all the objects stay in their original positions like on the picture:

![001](https://github.com/Scirra/Construct-bugs/assets/41597700/d6c2e7f6-fa01-4f76-b856-599239c26f28)
 

## More details

**Affected iPhones:**
XS Max

**Affected browsers/platforms:** 

**First affected release:** 
R350  
R370

## System details

<details><summary>View details</summary>

Platform information
Product: Construct 3 r370 (beta)
Browser: Chrome 119.0.6045.200
Browser engine: Chromium
Context: browser
Operating system: Windows 10
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 8
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36
Language setting: en-US

Local storage
Storage quota (approx): 71 gb
Storage usage (approx): 101 mb (0.1%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: no
Supports highp precision: yes
Vendor: Google Inc. (NVIDIA)
Renderer: ANGLE (NVIDIA, NVIDIA GeForce GTX 1070 (0x00001BE1) Direct3D11 vs_5_0 ps_5_0, D3D11)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_color_buffer_half_float
EXT_disjoint_timer_query_webgl2
EXT_float_blend
EXT_texture_compression_bptc
EXT_texture_compression_rgtc
EXT_texture_filter_anisotropic
EXT_texture_norm16
KHR_parallel_shader_compile
OES_draw_buffers_indexed
OES_texture_float_linear
OVR_multiview2
WEBGL_clip_cull_distance
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
WEBGL_multi_draw
WEBGL_provoking_vertex
Audio information
System sample rate: 48000 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

WebM Opus (audio/webm; codecs=opus)
Ogg Opus (audio/ogg; codecs=opus)
WebM Vorbis (audio/webm; codecs=vorbis)
Ogg Vorbis (audio/ogg; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM AV1 (video/webm; codecs=av01.0.00M.08)
MP4 AV1 (video/mp4; codecs=av01.0.00M.08)
WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
H.265 (video/mp4; codecs=hev1.1.2.L93.B0)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM AV1 (video/webm; codecs=av1)
WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

**Comment by Tarek200** [comment]
> Ugh, the iOS viewport handling is such a buggy mess. We have a bunch of workarounds for stuff like this already. I'd advise reporting an issue to Apple directly as well, as ideally this should all work on Apple's side and not rely on a pile of inconsistent workarounds on our side.

I totally agree, 
Apple is a nightmare to work with, is the only company that gives us so much stress. Every iOS update there are game-breaker bugs to the point that it becomes ridiculous. Causality is the company that you have to spend lots of money to work with, buying devices, Mac and Developers license 99 yearly, and they cannot even afford to make sure each iOS update is stable with all the money they make, which is ridiculous. We honestly had enough of working under this stress as we targeted Apple as the main platform and seeing all these bugs on each iOS is discouraging especially when the games take years to make.
I will try to open a bug with them swell, though I fear they will ask me technical questions about the engine that I will not be able to answer.

**Comment by Tarek200** [comment]
I quick update about the bug:

I noticed that it breaks when you open the app to resume the Game sometimes it forces a Game restart from the beginning like going through the loading screen etc... is when the canvas shifts up. So something happens that triggers the game to restart and is there between that restart that the issue happens. Also, I noticed that there is no limit to how far it can go up as each time it happens the canvas will go a bit farther up than last time.

![9](https://github.com/Scirra/Construct-bugs/assets/41597700/6ca17d6f-22b6-403c-9632-b46b45b7c4ed)


I checked other construct games if they have this problem and so far all the ones I tested have it:
For Example:

**Punch Golf:**

![4](https://github.com/Scirra/Construct-bugs/assets/41597700/f4578d9b-8410-4d45-9481-2d6dc24fef07)



![5](https://github.com/Scirra/Construct-bugs/assets/41597700/2d9a2d83-436a-4d5c-81f2-d875a931c4de)




**Little Dog Bob:**
![6](https://github.com/Scirra/Construct-bugs/assets/41597700/d7daa62f-642b-4fce-92f6-6726e62baa3b)



![7](https://github.com/Scirra/Construct-bugs/assets/41597700/a2d65198-eeaa-4233-b678-d4b9833dcce4)



![8](https://github.com/Scirra/Construct-bugs/assets/41597700/c9b8ba9c-2973-49d1-a1f3-c60cb1ee60f8)




**The Demo test I put on the Bug Report:**
![1](https://github.com/Scirra/Construct-bugs/assets/41597700/0ee385de-4615-43f1-8a0b-7c4ab1501ad1)



![2](https://github.com/Scirra/Construct-bugs/assets/41597700/dd3e4705-3bab-4afd-8e5f-3baa3ffd75c7)



![3](https://github.com/Scirra/Construct-bugs/assets/41597700/dcb2a9e1-d9e4-4b29-9f9a-e97e75feee36)

**Comment by Tarek200** [comment]
Also, I forgot to add that I updated to (iOS 17) and the bug still persists.

**Comment by therealPaulPlay** [comment]
I can replicate it and am pretty sure it is related to ram management. When I load up punchgolf and then play ram intensive games (asphalt, my construct game) and then return back, it shifts the canvas up. Probably some part of it is being unloaded. Closing and restarting the app fixes it for me (no reboot required).

**Comment by AshleyScirra** [comment]
I can't reproduce this. The provided project doesn't actually export out of the box because it is missing a bunch of project properties that need to be set for an export (e.g. description and ID). If I fill those out then it still doesn't work in Xcode due to some pods error. I also noticed the provided project is using a bunch of legacy compatibility settings (file: protocol, flat file structure mode) which modern projects don't use, and it is also not configured to use the full display (viewport fit is "Auto", not "Cover"). Are you sure you attached the right project?

I tried exporting the 'Flying along' template to Xcode as that's my go-to test project for mobile, which exported and ran fine. However after switching between apps repeatedly it never seemed to exhibit the issue described in this report. It is possible that certain project settings affect this issue, but it's difficult to tell.

To make further progress on this I'd need an up-to-date working minimal project that demonstrates the issue, and more reliable steps to reproduce. It may also depend on specific project settings, in which case knowing which settings affect it could be an important thing to know.

**Comment by therealPaulPlay** [comment]
I can at least say that my game is not affected by this, but it’s in landscape mode and uses scale inner.

**Comment by therealPaulPlay** [comment]
So I've tried the project myself (it does export just fine for me with the details filled in + file scheme set back to modern, installing the pods manually is necessary for me in each project on all my devices, even with an empty project and @Tarek200 seems to be facing the same issue (related to bug #7400 that is unfortunately still present). 

The bug does occur for me but only once, then it doesn't refresh the app anymore when doing the exact same thing and rather shows a black screen (and I need to restart my iphone to get it to move the canvas up again.)
I've also attached the fixed ios project with pods "installed" (although it's just a dependency issue).
![IMG_7688](https://github.com/Scirra/Construct-bugs/assets/80539587/e7c907e1-e173-497a-b4a5-6c109de0bc5a)
https://github.com/Scirra/Construct-bugs/assets/80539587/668cedd6-78bd-4b17-b56c-60fd110feed7


[New project.ios.project.zip](https://github.com/Scirra/Construct-bugs/files/13707292/New.project.ios.project.zip)

**Comment by Tarek200** [comment]
@therealPaulPlay thanks for your help I really appreciate it. Nice to see that there still are some mobile devs caring about their games :)) 


> "I can't reproduce this. The provided project doesn't actually export out of the box because it is missing a bunch of project properties that need to be set for an export (e.g. description and ID). If I fill those out then it still doesn't work in Xcode due to some pods error (I don't know why, but it doesn't seem related to the problem described here)."

The issue with pods started on Beta r371 as it was working fine on r368.2 but as you mentioned is not related to the bug or my settings.
**To fix this I have to do these steps:**
Terminal:
cd "Path Project folder"
pod deintegrate
pod install

**Source:**
https://stackoverflow.com/questions/55558984/xcode-pods-projectname-debug-xcconfig-unable-to-open-file-wrong-directory

![002](https://github.com/Scirra/Construct-bugs/assets/41597700/c26e2357-16ea-4b9e-8224-e3f54f32ced7)

---------------------------------------------

> "I also noticed the provided project is using a bunch of legacy compatibility settings (file: protocol, flat file structure mode) which modern projects don't use"


I use "Legacy" settings because "Moderm" gives a black screen as therealPaulPlay mentioned, though for me doesn't work at all as it starts with a black screen right away, so I thought that the "legacy" setting was meant to fix that. I think we need an up-to-date tutorial of what are the recommended settings to use because if you say that we cannot use legacy then the "Modern" setting option looks like has a bug. Also, maybe you could remove the legacy option if it's not meant to be used unless you have other reasons to leave it. 
Is a 3-year-old project which is why I use "Flat file structure mode" to avoid having to change all the "Load From URL" references.
I'm not sure I understood properly, are you saying that we cannot use "flat file structure mode" any more?

---------------------------------------------


> "and it is also not configured to use the full display (viewport fit is "Auto", not "Cover"). Are you sure you attached the right project?"

Yes, that's correct, that's the settings we want to use as we want to cut out the part of the notch and not cover the screen as it looks weird in our game at least.
Here is an example of what we want, that grey area is supposed to be cut out in black colour so the Top-Viewport starts below the Notch. also, it's much easier to set all the UI without having to check the safe areas.

![target - Copy](https://github.com/Scirra/Construct-bugs/assets/41597700/fddd3fdc-bed0-4df0-b015-d6ffa4b24307)


---------------------------------------------


> "To make further progress on this I'd need an up-to-date working minimal project that demonstrates the issue, and more reliable steps to reproduce. It may also depend on specific project settings, in which case knowing which settings affect it could be an important thing to know."

Not sure what else I can give as I posted a minimal project and also I linked two other c3 games that are currently on the store that have the same issue which should help you to test. 
@therealPaulPlay example should be good swell, though I dont think this is related to any settings or to Viewport fit (Cover & Auto).

**Comment by therealPaulPlay** [comment]
Just for clarification, I‘ve simply attached the xcode project of your c3p with the pods installed and the details filled in, I actually couldn’t get it past the cordova loading screen on ios with file structure set to legacy. Anyway, I hope this will be fixed :)

**Comment by Tarek200** [comment]
> Just for clarification, I‘ve simply attached the xcode project of your c3p with the pods installed and the details filled in, I actually couldn’t get it past the cordova loading screen on ios with file structure set to legacy. Anyway, I hope this will be fixed :)

Yep, iOS export is a mess as each export is a new world :)

But I found a way to reproduce it 100%
Also, is using the recommended setting that Ashley mentioned so the issue happens with both (Modern & Legacy)

**Here are my settings:**
-Cordova iOS scheme = Modern
-Export Files Structure = Folders
-Viewport Fit = Auto  (As we want to cut the notch area)
-ID = I put my temporary but I dont think you can use it as I believe they are unique like the domain websites so most probably you will have to replace it with yours in Xcode

**Ashley: Just so we are on the same page:**
This is our target design
![010](https://github.com/Scirra/Construct-bugs/assets/41597700/aaef5a07-9009-4a76-9696-eec422d55b6f)

And this is how it looks when it brakes
![012](https://github.com/Scirra/Construct-bugs/assets/41597700/0d60cfec-a574-491b-aae7-f420bd3ee6b0)


 


**Steps to reproduce:**
1-Donwload the C3 game "Little Dog Bob" from the iOS store and dont open it yet after it finishes downloading.

2-Export my project as iOS Cordova

3-Install it with Xcode (I noticed if you use the Avobe settings doesn't have the issue of Pod)

4-Open my App and let it load full 

5-Then close it

6-Open the "Little Dog Bob" ----> Click Play ----> enter the level and play for a few seconds

7-Close the App 

8-Open my App Again  

**Observed result:**
1-It refreshes the App 
2-Then loads it
3-But the canvas shifted up 
4-Then it goes to a black screen (With the previous C3 Versions it didn't happen that much the black screen not sure why now goes too often)  

**Note:**
Sometimes it refreshes too quickly that it just goes straight to a black screen.


**Here is a video of all the Avobe steps:**

https://github.com/Scirra/Construct-bugs/assets/41597700/6c567b2b-c6bb-4bcc-ae58-be4aa62f75dc

**Comment by therealPaulPlay** [comment]
Yeah well, Xcode probably is the biggest mess lol. As a workaround you could use viewport cover + fullscreen scale inner/outer and then offset the ui objects using platforminfo.SafeAreaInsetTop

**Comment by AshleyScirra** [comment]
> The issue with pods started on Beta r371 as it was working fine on r368.2

Can you identify the *specific* release the issue started happening in, if this is a regression? With difficult issues this can be really helpful information. We need to know the very last release it worked in and the very first release it stopped working in, including all beta releases - the wider the range the less helpful.

**Comment by Tarek200** [comment]
> > The issue with pods started on Beta r371 as it was working fine on r368.2
> 
> Can you identify the _specific_ release the issue started happening in, if this is a regression? With difficult issues this can be really helpful information. We need to know the very last release it worked in and the very first release it stopped working in, including all beta releases - the wider the range the less helpful.
> 
> Please note it's also important to have concise information. This is now a very long thread with lots of information. That can actually make it harder to deal with the problem. Ideally we need a small amount of the most precise and helpful information possible, so if anyone can write such a summary, it would be useful.

Afaik the Pods issue is a separate problem as it affects only the installation with Xcode, example it will be (succeed or fail) the installation unless you do the extra steps:
pod disintegrate
pod install

Depends on which settings you use (Legacy or Modern) 
I just gave you more detailed info because you asked for it in your last comment as your reference that you had issues installing the App with Xcode. 
I think we can leave that issue or open another report for that exclusively.
Not sure if Pod is related to my bug thought but I will leave it aside and I will just make a summary of my original issue (Canvas shift Up)
I will try to identify the first C3 version that causes the issue but please you guys on your team need to focus on this bug and help us as this is a very long process to debug, if possible we need to fix this before you take your Christmas holidays.

**Comment by therealPaulPlay** [comment]
I think we need to break it down into these 3 individual issues:
1) modern file scheme results in black screen on app refresh (specific to this project)
2) content moves up on background app refresh with fullscreen letterbox scale, portrait orient. and viewport fit auto
3) empty podfile generating and pods not installing (for me. That has been the case for almost a year already with every project)

**Comment by therealPaulPlay** [comment]
> > The issue with pods started on Beta r371 as it was working fine on r368.2
> 
> Can you identify the _specific_ release the issue started happening in, if this is a regression? With difficult issues this can be really helpful information. We need to know the very last release it worked in and the very first release it stopped working in, including all beta releases - the wider the range the less helpful.
> 
> Please note it's also important to have concise information. This is now a very long thread with lots of information. That can actually make it harder to deal with the problem. Ideally we need a small amount of the most precise and helpful information possible, so if anyone can write such a summary, it would be useful.

I‘ll file a new issue for the pods later when I’m home and do some research. At least for me it‘s not top priority though, pod install takes like 5s but especially for people who don‘t know about this bug and don’t have cocoapods installed it can be quite a hurdle.

**Comment by Tarek200** [comment]
> I think we need to break it down into these 3 individual issues:
> 
> 1. modern file scheme results in black screen on app refresh (specific to this project)
> 2. content moves up on background app refresh with fullscreen letterbox scale, portrait orient. and viewport fit auto
> 3. empty podfile generating and pods not installing (for me. That has been the case for almost a year already with every project)

Thats a good break down, I will just add that on Point 2:
Using scale Outer ----> it brakes too, that's the one I use on the examples I posted here  
Viewport Fit = Cover ----->  brakes too, I know this from the "Punch Golf" game that I posted above as starts the game on Cover mode and then when it brakes the canvas shifts up.

**Comment by Tarek200** [comment]
Something changed between the: 
R272 
& 
R272.2

The steps that I posted in my last comment to reproduce the bug anytime, are not working on 272.2 but still working on 272.
Is there any change on 272.2? 

Here is the Xcode console log when it refreshes the App and goes to a black screen, using R272:

Apache Cordova iOS platform version 7.0.1 is starting.
The preference key "AllowNewWindows" is not defined and will default to "FALSE"
The preference key "MediaPlaybackAllowsAirPlay" is not defined and will default to "TRUE"
The preference key "LimitsNavigationsToAppBoundDomains" is not defined and will default to "FALSE"
The preference key "AllowBackForwardNavigationGestures" is not defined and will default to "FALSE"
The preference key "Allow3DTouchLinkPreview" is not defined and will default to "TRUE"
CDVWebViewEngine will reload WKWebView if required on resume
Using WKWebView
[CDVTimer][console] 0.0319ms
[CDVTimer][handleopenurl] 0.0130ms
[CDVTimer][intentandnavigationfilter] 1.5010ms
[CDVTimer][gesturehandler] 0.0180ms
[CDVTimer][statusbar] 1.1611ms
[CDVTimer][TotalPluginStartup] 2.8859ms
THREAD WARNING: ['LaunchScreen'] took '172.9388' ms. Plugin should use a background thread.
[Construct] Suspending
CDVWebViewEngine shouldReloadWebView::
CDVWebViewEngine shouldReloadWebView title: iOS canvas Bug
CDVWebViewEngine shouldReloadWebView location: app://localhost/index.html
CDVWebViewEngine shouldReloadWebView reload: 0
[Construct] Resuming
[Construct] Suspending
CDVWebViewEngine shouldReloadWebView::
CDVWebViewEngine shouldReloadWebView title: iOS canvas Bug
CDVWebViewEngine shouldReloadWebView location: app://localhost/index.html
CDVWebViewEngine shouldReloadWebView reload: 0
[Construct] Resuming
[Construct] Suspending
CDVWebViewEngine shouldReloadWebView::
CDVWebViewEngine shouldReloadWebView title: iOS canvas Bug
CDVWebViewEngine shouldReloadWebView location: app://localhost/index.html
CDVWebViewEngine shouldReloadWebView reload: 0
[Construct] Resuming
[Construct] Suspending
CDVWebViewEngine shouldReloadWebView::
CDVWebViewEngine shouldReloadWebView title: iOS canvas Bug
CDVWebViewEngine shouldReloadWebView location: app://localhost/index.html
CDVWebViewEngine shouldReloadWebView reload: 0

That could be helpful if you changed something between these two versions I think?

**Comment by Tarek200** [comment]
> Something changed between the:
> R272
> &
> R272.2

Sorry I made a mistake as I should have referenced instead (R372 & R372.2) Sorry for the confusion, I will update my last comment

**Comment by AshleyScirra** [comment]
> The steps that I posted in my last comment to reproduce the bug anytime, are not working on 372.2 but still working on 372.
Is there any change on 372.2?

FWIW the sole change between r372 and r372.2 is a single bug fix for savegames. I do not think this would affect iOS exports in any way. I would guess the problem is intermittent and is incorrectly being attributed to things that don't actually affect it (which is pretty common).

**Comment by Tarek200** [comment]
Not sure why I can 100% use those steps to reproduce it on 372 but not on 372.2 as I use the same settings and the same way of installing the App with Xcode. Though that is just one of the ways to reproduce the bug as the bug is still present in all the C3 versions including 372.2 it just may take longer for the bug to appear, I have to mention this to make sure we are on the same page.
We have a problem now as we trying to give you specific steps to reproduce the bug but you are not able to reproduce it and the thread is getting longer so is better to go back to my original bug report which was simpler and broader, instead of targetting specific C3 versions or steps.   
So I think the best option is to start ober, I will make a new summary in my next comment.

**Comment by Tarek200** [comment]
We starting over you can forget all the comments above:

**New Summary:**
Here is my original report, as we cannot give you specific steps as it's all random but at least we know that it happens when you open the App while it is on Pause in the background.
![00gh1](https://github.com/Scirra/Construct-bugs/assets/41597700/7ce470bf-e773-4baa-a747-2072ac2bb4da)

**Observed Results:**
1. modern file scheme results in a black screen on app refresh
2. Canvas moves up on background app refresh


**Tips:**
-One tip that helps to reproduce the bug quicker is to keep switching between this App and other installed Apps on your phone until it happens.

Example:
1. close this C3 App
2. open another different App do something then close it
3. open again the C3 App.

Keep repeating the same steps until you can reproduce the bug. 

Again I have to keep emphasising this point:
> The bug doesn't happen right away so you may try on the first try and you will not be able to reproduce it, so please keep trying for one or two or as many days as necessary as sometimes happens after it has been paused in the Background for hours. We dont know the exact way that this happens as it happens randomly.

**Comment by therealPaulPlay** [comment]
Just to add one thing, I've recently experienced the same bug with my games (with fullscreen scaling, scale inner) and it actually happens too, you can see it when you record a video of it and slow it down, but the viewport is being adjusted in the next frame (presumably by Construct) to match the screen again. Also I was only able to replicate this bug by switching to other apps that were also using the webview (I tried it with bigger games like Minecraft etc. but that didn't work, when I used a simple news app, that did work)

**Comment by AshleyScirra** [comment]
I've got a speculative fix for this in the next beta, working on the assumption another iOS alignment bug might now be happening on resume. These issues are always a nightmare though so not sure how it will work out, so flagging this issue for others to re-test when the next beta is out.

**Comment by AshleyScirra** [comment]
It may not be feasible for us to fix this if the problem is with Apple's software (which seems likely in this case). I would advise to file an issue with Apple if this has not been done already.

**Comment by Tarek200** [comment]
> Your last comment mentions a crash rather than a display misalignment. That is a different problem and needs to be filed separately, as mentioning it in the same issue as a different issue which is already complicated only makes it harder to solve problems and less likely it can be corrected. It is important issue reports focus on a single thing and that discussion remains relevant to that.

**But you recommended we use:**
Cordova iOS Scheme: Modern Apps
![00x9](https://github.com/Scirra/Construct-bugs/assets/41597700/ccb7b415-0503-4fb2-afa7-624653c2d675)
That's why we get the Black screen with those settings as we mentioned many times above.


**I was using instead in my project:** 
Cordova iOS Scheme: Legacy files
![0s08](https://github.com/Scirra/Construct-bugs/assets/41597700/0b6a6940-a36d-45c6-b292-99a78b12fa0f)

Which doesn't give a black screen but has the issue of the "Canvas shifting up"

I would like to keep using "**Legacy files**" then but the latest r374 doesn't work anymore as when the App is installed it doesn't go past the loading screen, it gets stuck there so I cannot test if the fix you did is working or not.
![IMG_4572](https://github.com/Scirra/Construct-bugs/assets/41597700/3c736d07-6fb4-473a-bbfe-c2cd1483488d)

What do we need to do now, is that another bug to report separately?

**Comment by therealPaulPlay** [comment]
The one with the cordova scheme and black screen doesn‘t seem related to the canvas shifting up so I‘d file a separate bug report for that. I can replicate the issue with your project btw but haven’t experienced in any of mine.

**Comment by Tarek200** [comment]
We decided to give up on iOS and move on as we spent around two months trying to fix these bugs, using time that we dont really have, working for free and it's just going around in circles. We pay C3 so we dont have to deal with third-party companies, this is not our job to report to Apple. There are so many issues and bugs here anyway, each new version has a new bug, this is not stable and is impossible to make something serious in the state that it is now.
We are very disappointed as we spent 3 years full-time on this game and our main target was iOS, we built everything around it so having to deal with all these bugs at the end when the game is finished is crazy and caused us way too much stress than it's worth.
With each new iOS there are game-breaking bugs, I dont understand why you keep using Cordova technology when it's been very unstable for so many years.

If anyone wants to take over the bugs go ahead, I'm done with them.

**Comment by Tarek200** [comment]
Hey @therealPaulPlay  Thanks for your help I appreciate your time, Good luck with your Game :)

**Comment by therealPaulPlay** [comment]
> We decided to give up on iOS and move on as we spent around two months trying to fix these bugs, using time that we dont really have, working for free and it's just going around in circles. We pay C3 so we dont have to deal with third-party companies, this is not our job to report to Apple. There are so many issues and bugs here anyway, each new version has a new bug, this is not stable and is impossible to make something serious in the state that it is now. We are very disappointed as we spent 3 years full-time on this game and our main target was iOS, we built everything around it so having to deal with all these bugs at the end when the game is finished is crazy and caused us way too much stress than it's worth. With each new iOS there are game-breaking bugs, I dont understand why you keep using Cordova technology when it's been very unstable for so many years.
> 
> If anyone wants to take over the bugs go ahead, I'm done with them.

I can totally agree that it is quite a struggle and I‘m having a lot of issues as well, but these two seem pretty fixable at least. You could for example use fullscreen scaling and then just get the upper screen offset for devices using a notch with a plugin. I‘d recommend opening a separate report for the black screen issue if necessary. 

The state of ios performance is not the best to be honest, but I can say that (for some reason) web gl performance is miles better on newer ios devices, although the gpu performance diff shouldn‘t be too noticeable. Also, I‘m hoping for web gpu to fix some of the odd rendering issues and related crashes in webkit, though it might as well introduce a bunch of new ones. I‘m also using a bunch of whacky work-arounds and plugins to make everything work but to be honest, that’s necessary with most engines (except unity…). 

I‘m happy that android performance is at least good, though some devices with higher refresh screens only render everything properly and without hiccups or random slow downs with latency set to low which breaks the preview and that doesn‘t seem to get fixed :(

Lastly I‘d encourage you to make use of cordova plugins, there are many useful ones that fix (somewhat) broken parts of cordova (such as loading).

**Comment by AshleyScirra** [comment]
I've got another speculative fix in for the next beta so it would be worth checking that again when it's out. If that does not help I'm afraid I don't think there's anything more we can do, so you'll have to report it to Apple.

I've tried my best to do what we can but as our bug report guidelines note, sometimes the problem is with somebody else's software and so we cannot fix it and reporting it to us will not help, and so our standard procedure is to recommend that you file an issue with the affected software vendor directly (in this case with Apple). When we have difficultly reproducing the problem ourselves then having the customer file the issue themselves can be the only viable way to make progress on the problem. It might be annoying but I'm afraid that's just the way things work in software sometimes. We don't control everything that happens in the software world and we can't fix everything that goes wrong in other people's software.

**Comment by AshleyScirra** [comment]
Please try again with r375 - if we're lucky that might have helped...

**Comment by Tarek200** [comment]
Thanks
I managed to get around the black screen and test your fix on the r375 but the issue persists.


One thing I noticed is once it triggers the App Reload it will break every C3 game that you open as it forces each C3 App to Reload when you open it so they end up all with the "Canvas moved Up".



![Canvas Bug](https://github.com/Scirra/Construct-bugs/assets/41597700/db3be1ad-92ed-4aa2-a893-bc26714aaa1b)
![Canvas Bug2](https://github.com/Scirra/Construct-bugs/assets/41597700/674c9e32-67a5-4d7c-bb44-40a31ef79fba)

I'm not sure why it doesn't happen with other games made for example with Unity using web teach.

We will have to pass from iOS as we dont have time for this and we will focus on other platforms like (Steam, Android, etc...) and maybe try iOS again in the future to see if these issues have been fixed. 
Even if we report it to Apple it could take months or years to fix, which is time we dont have as we are just a small indie team and we dont have the resources to keep waiting for a fix. The last iOS 15 issues took between 6 to 12+ months to get resolved so it's a bit unrealistic for us to keep following these bugs to Apple or third parties even if we want to.

But Thanks for trying.

I'm amazed how no other users are reporting this issue as all the C3 games that I tested are affected.

**Comment by AshleyScirra** [comment]
Sorry folks, this one will have to be reported to Apple if you're affected. I've tried adding workarounds that make sure it updates to the correct window size and resets the scroll position after waking, and apparently neither have had any effect. I don't know what might be causing a display offset; perhaps it's in the app view itself, or the system. It might be a bug in Cordova iOS, or in iOS itself - either way it's not something in our control, and given the difficulty we've had reproducing it, it's best affected customers file issues themselves directly (which is sometimes our standard practice as per the bug report guidelines).

If the app is crashing that would be a different issue, and would probably also need to be reported to Apple directly, as even malfunctioning JavaScript code does not normally cause an app crash (system-level issues like graphics driver bugs are normally the cause of such issues).

**Comment by therealPaulPlay** [comment]
Thanks for your efforts. I have tried this with my version of the test project and I haven't noticed any changes in the behavior.

Because of the black screen / not reloading issue I cannot test it properly with the Modern Cordova scheme and the legacy scheme is broken for me in r375. I'm not sure how @Tarek200 you were able to work around the black screen issue (you said that you found a workaround for that). 

I've filed a new bug report for the reloading bug and have narrowed it down to version r370 in which it broke for me. This is a very important issue as it completely breaks most games in many scenarios, not only when opening a browser, also sometimes when reopening the game or when switching layouts etc., so please please address this #7748

I've also filed a bug for the legacy scheme not going past the cordova splash screen on iOS but I don't think it is high priority as I don't see a real reason to use it apart from the former bug with the modern scheme. #7749

By doing close to 40 exports in different browsers I've finally found the reason for the pod file issue and have filed a new issue regarding that topic: #7750 This is also not that important to me because I have to install the pods manually anyway as I'm using 3rd party plugins as discussed here: #6934

**Comment by Tarek200** [comment]
> Because of the black screen / not reloading issue I cannot test it properly with the Modern Cordova scheme and the legacy scheme is broken for me in r375.
> 
> Thanks I managed to get around the black screen and test your fix on the r375 but the issue persists.
> 
> I'm not sure how @Tarek200 you were able to work around the black screen issue (you said that you found a workaround for that).

The (black screen & Freeze Splash Screen) issues are caused by the same source problem so they could fix both bugs by resolving the origin. 

**Here is how I fixed both problems (black screen & Freeze in Splash Screen)**

**1-The Problem:**
I found out that the issue happens when you upgrade a project from an old C3 version to a newer version you will get the (black screen & Freeze in Splash Screen), I think it could be because maybe it leaves some old files or references in the new C3 version that you upgraded to which conflicts with Cordova? 
Not sure but Ahsley has to confirm that. I believe this because of how I managed to solve it.

**2-The Solution:**
I managed to resolve this (black screen & Freeze in Splash Screen) problem by recreating the Demo that I added to this bug report using the last r368-2. So I created a new project on the r368-2 and I cloned all the events, settings and assets. So basically a clone of the project but done directly on the r368-2 version.

**So you have:**
The Demo project that I added to this bug report was originally done on the R218  so now that has been upgraded to the latest r375 and exported the Cordova file and test you will get:

-If you use Cordova iOS Scheme: Modern Apps  -----> Results in (Black screen)
-If you use Cordova iOS Scheme: Legacy Apps  -----> Results (Freeze in Splash Screen)

(The reason why I went that far to R218 is because I wanted to find out in which C3 version started the issue of the canvas shifting up so you have to start from low and then keep moving up versions), I'm just putting it out there because I'm sure someone will ask.

**Note:**
If this is true could be a big problem for devs that upgrade from the old C3 version to a new one, For example, my game was originally done on C2 and then I upgraded around 3 years ago so projects coming from an old c3 version could have many issues like the ones mentioned above, when using iOS exports.
Also, the fix I did by recreating the demo on the latest r375 should be counted as a just temporary fix because is a small demo but if you have to recreate a 3 years game project, it is unrealistic so the best solution is to find out what (references or files) is causing the problem when you upgrade to a newer C3 version and remove them.  


Here is the New Demo that I did on r368-2 which doesn't have the (black screen & Freeze Splash Screen) issues: (Just open it on the latest r375 and export with Cordova)
Also, you can use this for regarding the main bug of this thread, the one that moves up the canvas, for opening the ticket on Apple. 

> Regarding the main bug of this thread, the one that moves up the canvas, I'd file a bug with apple but unfortunately as explained I cannot create a viable test project because of the reloading/blackscreen bug

[iOS Canavs BUG.zip](https://github.com/Scirra/Construct-bugs/files/13976210/iOS.Canavs.BUG.zip)

**Comment by therealPaulPlay** [comment]
Thanks for your reply. Unfortunately, when I create a project in a newer version (like 375) the Blackscreen also happens, not only when I upgrade existing (older) projects. (Regarding the faulty webview reload). Can't speak for the Legacy Scheme bug though, I have indeed tested this with a project created in an earlier version which I then updated as far as I remember. 

I have a huge project as well (like 20k actions) and everything is working fine in r368.2, but when I upgrade to r370 (though I assume 369 introduced these bugs) the game doesn't even load the first layout as it is getting stumped on the first reload that is being triggered. When I create smaller projects in r375, loading times also appear to be a lot longer and as explained in the bug report, the reload often times fails which leads to the app being completely unresponsive

**Comment by Tarek200** [comment]
> Thanks for your reply. Unfortunately, when I create a project in a newer version (like 375) the Blackscreen also happens, not only when I upgrade existing (older) projects. (Regarding the faulty webview reload). Can't speak for the Legacy Scheme bug though, I have indeed tested this with a project created in an earlier version which I then updated as far as I remember.
> 
> I have a huge project as well (like 20k actions) and everything is working fine in r368.2, but when I upgrade to r370 (though I assume 369 introduced these bugs) the game doesn't even load the first layout as it is getting stumped on the first reload that is being triggered. When I create smaller projects in r375, loading times also appear to be a lot longer and as explained in the bug report, the reload often times fails which leads to the app being completely unresponsive.
> 

Interesting, if the problem is not by how old the version then it looks like it may be just specific to a version that (Added or Changed) something. At least we can confirm that r368.2 is working great for The (black screen & Freeze Splash Screen) issues and the issues start after that r368.2 version.

> Interestingly, my project is unaffected by the canvas shift bug, likely because I'm using viewport cover. I've also noticed that whenever you open a safari window for example, it shifts the canvas back to the correct position in the c3 app, I'm not sure whether that is connected to the bug.

Honestly, I will recommend you to keep double-checking as all the c3 I tested have the canvas issue. Also, on my latest test with r375, it took me more than 1 Hour to reproduce the canvas bug for the first time, I had been switching between Apps not necessarily between just C3 games and when it triggered the Reload all the C3 games I open it (Reloaded and braked the Canvas).
If you share the link to your Game I can test it and see if it breaks for me. 
If you dont want to share it publicly you can send it privately on Discord or any other platform.


> Interestingly, my project is unaffected by the canvas shift bug, likely because I'm using viewport cover

I dont think any settings are safe from the canvas bug, if you download the game that I linked on this thread "Punch Golf" you can see that he uses  "Viewport Cover" but I managed to reproduce the bug so many times on that game, so all the settings are affected (Cover, Auto, Scale inner, Scale Outer, etc...

![Punch Golf](https://github.com/Scirra/Construct-bugs/assets/41597700/33773f10-d9b0-46fd-8cba-6bb69a9f5883)

**Comment by therealPaulPlay** [comment]
> Honestly, I will recommend you to keep double-checking as all the c3 I tested have the canvas issue. Also, on my latest test with r375, it took me more than 1 Hour to reproduce the canvas bug for the first time, I had been switching between Apps not necessarily between just C3 games and when it triggered the Reload all the C3 games I opened it (Reloaded and braked the Canvas). If you share the link to your Game I can test it and see if it breaks for me. If you dont want to share it publicly you can send it privately on Discord or any other platform.
> 
> > Interestingly, my project is unaffected by the canvas shift bug, likely because I'm using viewport cover
> 
> I dont think any settings are safe from the canvas bug, if you download the game that I linked on this thread "Punch Golf" you can see that he uses "Viewport Cover" but I managed to reproduce the bug so many times on that game, so all the settings are affected (Cover, Auto, Scale inner, Scale Outer, etc...

I've been able to reproduce the bug with your example project, with punch golf as well as little dog bob. However, I have spent at least 30+ hours testing my game on iOS and am 100% sure that it is unaffected. After doing some tests with viewport cover as well I can confirm that it is affected too. Maybe the reason that it does not occur for me is because I'm using this [addon ](https://www.construct.net/en/make-games/addons/927/hide-homebutton-ios) that I've created for proper iOS fullscreen or because I'm using "hide statusbar". I think the latter seems more likely, though I'm not sure whether you have tested that already.

**Comment by fredriksthlm** [comment]
Try to replace the C3 Index HTML file with a "blank" HTML (maybe just with a specified BG color). Build that app, and test it just like you have done above.
If you still see the issue anytime then this issue is not related to Construct.

**Comment by Tarek200** [comment]
> I've been able to reproduce the bug with your example project, with punch golf as well as little dog bob. However, I have spent at least 30+ hours testing my game on iOS and am 100% sure that it is unaffected.

I see, though we have different ways of testing so you never know I may be able to replicate it )) But anyway even if I was able to reproduce it with your game is not gonna fix or help on anything.

> Maybe the reason that it does not occur for me is because I'm using this [addon ](https://www.construct.net/en/make-games/addons/927/hide-homebutton-ios)that I've created for proper iOS fullscreen or because I'm using "hide statusbar". I think the latter seems more likely, though I'm not sure whether you have tested that already.

I don't think is "hide statusbar" as I always use it on my games.
About your "addon" we will not know unless more people can test your game and see if it brakes or not but if its true that it doesn't then it may have the key for a solution.   

> I've also noticed that whenever you open a safari window for example, it shifts the canvas back to the correct position in the c3 app, I'm not sure whether that is connected to the bug.

I forgot to mention that I did a test to see if this canvas bug is caused by only switching between Apps, so I left my demo App untouched the whole night and we didn't open any other game so today when we opened it, it did reload and brake the canvas on my test demo, the only thing that we used on the night was the safari so that can be another source of triggering the reload or it maybe just because it has then been too long on pause that triggered the reload. All it seems once it triggers that reload it breaks the games.

**Comment by therealPaulPlay** [comment]
Safari is another source of triggering this, I have already confirmed that. Opening a new tab is like opening a construct application in that regard. @Tarek200 if you want to test my app your way (though I think we test the apps similarily) I can send you an old build on discord (if you share your dc user name). @fredriksthlm that's a really good idea, I'm unfortunately very busy today so I can't test that rn

**Comment by Tarek200** [comment]
>  @Tarek200 if you want to test my app your way (though I think we test the apps similarily) I can send you an old build on discord (if you share your dc user name)

What I meant was the link to your Game in the store so I can download it and test it, not the exported project. Though only share it if you want to make sure if it's broken, else it's not necessary as we have many examples here already.

**Here is another C3 game broken:**
![Layer 0](https://github.com/Scirra/Construct-bugs/assets/41597700/41f836d4-ce83-45b6-a0bd-bf16d7dca551)

![Layer2](https://github.com/Scirra/Construct-bugs/assets/41597700/bb38614d-14cb-437f-b7a7-b0c4937234cc)

**Comment by Tarek200** [comment]
Another thing that I just noticed is that C3 is not consistent when using (ViewPort Fit = Auto) 

For example:

**On R368.2**
Look at the notch at the bottom it cuts the notch area

![r368](https://github.com/Scirra/Construct-bugs/assets/41597700/b5e807b3-2dd2-46c2-b365-5535c0776853)

On R369
Look at the notch at the bottom it covers the whole bottom screen

![r369](https://github.com/Scirra/Construct-bugs/assets/41597700/cef70aac-aa95-43d0-aea6-7b8c8ff83437)


====================================


**Could this be what is causing the Canvas OffSet issue?**

At the least, that's another bug to report separately as it has to be consistent on every version.

Honestly, iOS is totally broken, I can't believe it is in this state (:

The question now is which of the two pictures is the correct view

**Comment by therealPaulPlay** [comment]
My game has not been published yet, though as I said I'm sure that it is unaffected. If you want to see for yourself, I can send you an xcode export via discord... Really weird that it's also changing the canvas width this way

**Comment by therealPaulPlay** [comment]
Just something that I've noticed... the google admob window did not move up, so it's not the html window that changes place, it's the canvas2D, so likely related to construct (correct me if I'm wrong here)

**Comment by Tarek200** [comment]
> My game has not been published yet, though as I said I'm sure that it is unaffected. If you want to see for yourself, I can send you an xcode export via discord... Really weird that it's also changing the canvas width this way

I see, I thought your game was already in the store, no worries then.

> Just something that I've noticed... the google admob window did not move up, so it's not the html window that changes place, it's the canvas2D, so likely related to construct (correct me if I'm wrong here)

That's a good point, It was weird to see it stand in its original place, the Construct team could clarify if this is a hint regarding the canvas issue could be from C3, not Apple, as you would spect that the AD banner will move up swell if the canvas was broken from Apple.

**Comment by therealPaulPlay** [comment]
Well, I'm pretty sure it's an iOS related bug still, as the canvas2d does not move up on android (for example). It's just interesting that it's apparently related to the canvas and not the HTML page

**Comment by Tarek200** [comment]
Mm using:
r375
Worker Mode = Off
Scale Outer
A sprite rotating with Orbit behaviour

-I spent 4 hours yesterday and it didnt reload even once, I wasn't able to trigger the reload on this demo project when all the other C3 games installed reloaded like hundred times. Normally I could easily trigger the reload but not now using those especifications.

-However, after leaving it all night I opened it today again and it triggered the reload but it didn't go to a black screen so it just loaded normally and shifted the canvas up, though I observed something good when I closed the App by sending it to the background and I open it right away it fixed the Canvas offset without reloading the App.

The fix you did Ashley seems to have some effect using those specifications above but:
-It didn't fix it on App Reload as I had to close the App and open it again to see the OffSet has been resetting
-This just happened once as I could manage to reload only once so I need to do more testing.

**This is what happened when it reloaded the App:**
![Layer 0](https://github.com/Scirra/Construct-bugs/assets/41597700/380db32d-478a-4259-ab83-cb39e8dc2ea6)



**And this is what happened when I closed and open the App again:**
![Layer 0](https://github.com/Scirra/Construct-bugs/assets/41597700/b6d505eb-2490-4c45-a088-bd980075bd15)





======================================================================

======================================================================

======================================================================

 
**Also, another important observation that I saw is:**

Web-exported games don't have the canvas bug issue as I used Safari to install It as an App and I use the same steps to make it reload as I do with Cordova ios Apps and when it reloads the canvas doesn't break despite I managed to make it reload 100 times.

**Here is what I did to test this:**
-Export as an HTML
-Host the game on your website or local server.
-Open the link on Safari
-Press the share button and choose "Add to Home Screen"
-Then just open the App from the home screen and follow the steps to make it reload 

![Layer fg0](https://github.com/Scirra/Construct-bugs/assets/41597700/ae786b46-1460-44ed-ab16-c1643d3e69bd)


The question is shouldn't be both exports option broken? (Cordova ios & Web) if they use the same tech, not sure I just find it interesting.

**Comment by Tarek200** [comment]
I added a video into my last comment showing how the canvas breaks after triggering the reload and then after closing the App and reopening it it fixes the canvas Offset,

**Comment by therealPaulPlay** [comment]
Just fyi, I could get the canvas reload to happen in the web export by simply opening a lot of apps so that safari itself does this, which consequently does the same to the tab and has the same effect

**Comment by Tarek200** [comment]
> Just fyi, I could get the canvas reload to happen in the web export by simply opening a lot of apps so that safari itself does this, which consequently does the same to the tab and has the same effect

Do you mean that the canvas brakes on web exports for you? 
I never had this issue on web exports

The reason I save it as an app when I export it as HTML is to get the full screen so it runs like a normal iOS App. But yeah the web exports its much easier to trigger the reload.

**Comment by therealPaulPlay** [comment]
Yes it breaks, you just need to get Safari to reload instead of the construct app  or web page which is a lot harder… Same goes for web apps that you add to your home screen, because they ultimately just use safari

**Comment by Tarek200** [comment]
> Yes it breaks, you just need to get Safari to reload instead of the construct app or web page which is a lot harder… Same goes for web apps that you add to your home screen, because they ultimately just use safari

I see, 
I wasnt able to reproduce on web exports despite that the App triggered the reload hundreds of times, I guess is because we are using different settings. 

Here is a small video showing a web export added to my Home screen triggering the reloading but still maintaining the canvas in the right position but all the other C3 games are braking. The worst one is the doggy.


https://github.com/Scirra/Construct-bugs/assets/41597700/ee15c312-e286-4561-a41d-63ed8e2bbd68


Anyways I think the fix Ashley did is nearly there but it should be applied on App reload so the players don't see when the App braking at all. iOS is still very problematic as is to much time consuming I will skip it for now and just concentrate on other platforms.

**Comment by therealPaulPlay** [comment]
It‘s a lot harder to replicate in web exports as I said because the „reload“ that happened hundreds of times is not the same type of reload than the one that entire apps do. However, I‘m talking about r375 here, I couldn’t replicate it in r376 however I was able to get the canvas shift to happen on the construct loading screen (not in-game though, as the game starts, it’s going back to the correct position)

**Comment by therealPaulPlay** [comment]
At least for me, putting overflow: hidden on the entire body prevents safari from scrolling down on refresh, just putting this out there as a potential workaround. 
I've also found a possible cause for this issue (as seen in the bug report)
